lib/tom_queue/deferred_work_manager.rb

```diff
@@ -163,69 +63,50 @@ def setup_amqp(channel)
     # payload  - (String) the message payload
     #
     #
-    def thread_consumer_callback(response, headers, payload)
+    def schedule(response, headers, payload)
       run_at = Time.at(headers[:headers]['run_at'])
 
       # schedule it in the work set
-      @deferred_set.schedule(run_at, [response, headers, payload])
+      deferred_set.schedule(run_at, [response, headers, payload])
     rescue Exception => e
       r = TomQueue.exception_reporter
       r && r.notify(e)
-      
+
       ### Avoid tight spinning workers by not re-queueing redlivered messages more than once!
       response.channel.reject(response.delivery_tag, !response.redelivered?)
     end
 
-    # Internal: The main loop of the thread
-    #
-    def thread_main
-      debug "[thread_main] Deferred thread starting up"
-
-      # Make sure we're low priority!
-      Thread.current.priority = -10
-
-      # Create a dedicated channel, and ensure it's prefetch 
-      # means we'll empty the queue
-      @channel = TomQueue.bunny.create_channel
-      @channel.prefetch(0)
-
-      # Create an exchange and queue
-      _, queue = setup_amqp(@channel)
-
-      @deferred_set = DeferredWorkSet.new
-      
-      @out_manager = QueueManager.new(prefix)
+    def start
+      debug "[DeferredWorkManager] Deferred process starting up"
 
       # This block will get called-back for new messages
-      consumer = queue.subscribe(:ack => true, &method(:thread_consumer_callback)) 
+      @consumer = queue.subscribe(:ack => true, &method(:schedule))
 
       # This is the core event loop - we block on the deferred set to return messages
       # (which have been scheduled by the AMQP consumer). If a message is returned
       # then we re-publish the messages to our internal QueueManager and ack the deferred
       # message
-      until @thread_shutdown
-        
+      while true
         # This will block until work is ready to be returned, interrupt
         # or the 10-second timeout value.
-        response, headers, payload = @deferred_set.pop(2)
+        response, headers, payload = deferred_set.pop(2)
 
         if response
           headers[:headers].delete('run_at')
-          @out_manager.publish(payload, headers[:headers])
-          @channel.ack(response.delivery_tag)
+          out_manager.publish(payload, headers[:headers])
+          channel.ack(response.delivery_tag)
         end
       end
-
-      consumer.cancel
-
-    rescue
+    rescue Exception => e
+      error e
       reporter = TomQueue.exception_reporter
       reporter && reporter.notify($!)
+    end
 
-    ensure
-      @channel && @channel.close
-      @deferred_set = nil
-      @thread = nil      
+    def stop
```

Comment:
I'm not sure of the ordering of this, vs. the main loop, given you're in an exception handler. I tend to do the minimum possible work in such handlers.

I'd suggest setting a `@shutdown` flag to true, then calling `deferred_set && deferred_set.interrupt` to make the main-thread loop spin once and hit the `until @shutdown` to break out. 

Then the two `consumer.cancel` and `channel.close` can live after the loop itself has exited.